src/pulsar_producer.erl

```diff
@@ -61,106 +99,227 @@ callback_mode() -> [state_functions].
 
-send(Pid, Message) ->
-    gen_statem:cast(Pid, {send, Message}).
-
-send_sync(Pid, Message) ->
-    send_sync(Pid, Message, 5000).
+-spec send(gen_statem:server_ref(), [pulsar:message()]) -> ok.
+send(Pid, Messages) ->
+    From = undefined,
+    erlang:send(Pid, ?SEND_REQ(From, Messages)),
+    ok.
 
-send_sync(Pid, Message, Timeout) ->
-    gen_statem:call(Pid, {send, Message}, Timeout).
+-spec send_sync(gen_statem:server_ref(), [pulsar:message()]) ->
+          {ok, send_receipt()}
+        | {error, producer_connecting
+                | producer_disconnected
+                | term()}.
+send_sync(Pid, Messages) ->
+    send_sync(Pid, Messages, 5_000).
+
+-spec send_sync(gen_statem:server_ref(), [pulsar:message()], timeout()) ->
+          {ok, send_receipt()}
+        | {error, producer_connecting
+                | producer_disconnected
+                | term()}.
+send_sync(Pid, Messages, Timeout) ->
+    Caller = self(),
+    MRef = erlang:monitor(process, Pid),
+    erlang:send(Pid, ?SEND_REQ({Caller, MRef}, Messages)),
+    receive
+        {MRef, Response} ->
+            erlang:demonitor(MRef, [flush]),
+            Response;
+        {'DOWN', MRef, process, Pid, Reason} ->
+            error({producer_down, Reason})
+    after
+        Timeout ->
+            erlang:demonitor(MRef, [flush]),
+            receive
+                {MRef, Response} ->
+                    Response
+            after
+                0 ->
+                    error(timeout)
+            end
+    end.
 
 %%--------------------------------------------------------------------
-%% gen_server callback
+%% gen_statem callback
 %%--------------------------------------------------------------------
-init([PartitionTopic, Server, ProxyToBrokerUrl, ProducerOpts]) ->
+
+init([PartitionTopic, Server, ProxyToBrokerUrl, ProducerOpts0]) ->
+    process_flag(trap_exit, true),
     {Transport, BrokerServer} = pulsar_utils:parse_url(Server),
+    ProducerID = maps:get(producer_id, ProducerOpts0),
+    ReplayqCfg0 =
+        case maps:get(replayq_dir, ProducerOpts0, false) of
+            false ->
+                #{mem_only => true};
+            BaseDir ->
+                Dir = filename:join([BaseDir, PartitionTopic]),
+                SegBytes = maps:get(replayq_seg_bytes, ProducerOpts0, ?DEFAULT_REPLAYQ_SEG_BYTES),
+                Offload = maps:get(replayq_offload_mode, ProducerOpts0, false),
+                #{dir => Dir, seg_bytes => SegBytes, offload => Offload}
+        end,
+    MaxTotalBytes = maps:get(replayq_max_total_bytes, ProducerOpts0, ?DEFAULT_REPLAYQ_LIMIT),
+    MaxBatchBytes = maps:get(max_batch_bytes, ProducerOpts0, ?DEFAULT_MAX_BATCH_BYTES),
+    ReplayqCfg =
+        ReplayqCfg0#{ sizer => fun ?MODULE:queue_item_sizer/1
+                    , marshaller => fun ?MODULE:queue_item_marshaller/1
+                    , max_total_bytes => MaxTotalBytes
+                    },
+    Q = replayq:open(ReplayqCfg),
+    ProducerOpts =
```

Comment:
Just wondering: we put replayq inside ProducerOpts to avoid state record upgrade during code change?

Reply:
Yes, that was the original goal. 

But after [this comment](https://github.com/emqx/pulsar-client-erl/pull/39#discussion_r972128848), I think I'll have to migrate the record to a map, anyway :joy_cat: